File: c/problem_0009.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/**
 * @see https://projecteuler.net:
 *
 * Special Pythagorean triplet
 * Problem 9
 *
 * A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
 * a2 + b2 = c2
 *
 * For example, 32 + 42 = 9 + 16 = 25 = 52.
 *
 * There exists exactly one Pythagorean triplet for which a + b + c = 1000.
 * Find the product abc.
 */

int main(void)
{
    int i, j, k;

    for (i = 1000; i > 0; i--)
    {
        for (j = i - 1; j > 0; j--)
        {
            k = 1000 - i - j;
            {
                if (k > 0 && k * k + j * j == i * i)
                {
                    printf("Found Pythagorean triplet %i + %i + %i = 1000\n", k, j, i);
                    printf("The product of them is: %i\n", k * j * i);
                    exit(0);
                }
            }
        }
    }
    printf("Hmm... No Pythagorean Triplet found that sums to 0... Something went wrong!\n");
}

File: haskell/problem_0009.hs
{--------------------------------------------------------------------
<p>A Pythagorean triplet is a set of three natural numbers, $a \lt b \lt c$, for which,
\$$a^2 + b^2 = c^2.$$</p>
<p>For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.</p>
<p>There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.<br>Find the product $abc$.</p>
---------------------------------------------------------------------}

import EulerLib qualified as E

main = do
  -- Idea: form a tuple list (a, b, c) for all possible combinations,
  -- then filter by valid combinations (a^2 + b^2 = c^2)
  -- then filter by a+b+c = 1000
  let eulerProblem = 9
  let maxNr = 1000
  -- Build cartesian product of a,b,c, for each number a < b < c:
  -- using a list comprehension:
  let abc = [(a, b, c) | a <- [2 .. maxNr - 2], b <- [a + 1 .. maxNr - 1], c <- [b + 1 .. maxNr]]
  let validProds = filter (\(a, b, c) -> a * a + b * b == c * c && a + b + c == 1000) abc

  let solution = head $ map (\(a, b, c) -> a * b * c) validProds
  E.printProblemSolution eulerProblem (show solution ++ " for the tuple " ++ show (head validProds))
